### PR TITLE
[12.0][MIG] pos_tare w/ barcode

### PR DESCRIPTION
**Pull request by Fkawala**
Following up our discussion with @legalsylvain here #436 and there #447 Sylvain and I decided to split the work related to tare in two parts.

- The first part handles manual tare input (as described in #436) and barcode base tare input (as described in #447). 
- The second part, to be released afterwards, handles the tare barcode label printing from the POS UI (as described in #447).

This PR covers the first part. The second part will be delivered later next week.

### CODE REVIEW

**Comment by legalsylvain** [comment]
Hi @Fkawala. I'm currently reviewing this PR.
I'll do a PR against your, to change something.
kind regards.

**Comment by Fkawala** [comment]
> Hi @Fkawala. I'm currently reviewing this PR.
> I'll do a PR against your, to change something.
> kind regards.

Works for me!

**Comment by Fkawala** [comment]
@legalsylvain this PR is pending a formal test with an ESCPOS printer. I'll try to get that done tomorrow. In the meantime would you have some spare cycles to review the code? 

@vvrossem this PR is the base layer for a proper tare / gross weight handling in the POS. It adds a "Tare" button and includes @legalsylvain's nice rework of the scale screen (first introduced in #436). This PR features the gross weight / Tare / net weight display on receipt (works for web print and proxy print). This PR is meant to become a dependency for #470 that you've already reviewed a while ago. Do you think you'll have time to review this? If not would you be kind enough to bounce it back to someone else at CoopITEasy? Thank you in advance.

**Comment by Fkawala** [comment]
The CI tests fail b/c:

`odoo.exceptions.UserError: ("You try to install module 'pos_tare' that depends on module 'base_fontawesome'.\nBut the latter module is not available in your system.", '')`

I'm not sure how to fix that!?

**Comment by Fkawala** [comment]
Sneak peek at the UI changes:

image
image
image

File: pos_tare/static/src/js/screens.js [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        set_weight: function (gross_weight) {
+            this.gross_weight = gross_weight;
+            var net_weight = gross_weight - (this.tare || 0);
+            this.$('#container_weight_gross').text(
+                this.get_product_gross_weight_string());
+            this._super(net_weight);
+        },
+
+        order_product: function () {
+            // TODO Set a warning, if the value is incorrect;
```
Comment (vvrossem) [inline pos_tare/static/src/js/screens.js]: Sometimes I forget my todoes, maybe it's a similar case here :man_shrugging:

File: pos_tare/static/src/js/screens.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+        check_sanitize_value: function (input_name) {
+            var res = this.$(input_name)[0].value.replace(',', '.').trim();
+            if (isNaN(res)) {
+                this.$(input_name).css("background-color", "#F66");
+                return undefined;
+            }
+            this.$(input_name).css("background-color", "#FFF");
+            return parseFloat(res, 10);
+        },
```
Comment (vvrossem) [inline pos_tare/static/src/js/screens.js:132]: Nice!

File: pos_tare/static/src/js/screens.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                        " zero. Call for help unless you're perfectly" +
+                        " sure you are doing right."), wrong_product),
+                    confirm: function() {
+                        _super_validate_order();
+                    },
+                });
+                return;
+            }
+            return this._super(options);
+        },
```
Comment (vvrossem) [inline pos_tare/static/src/js/screens.js:158]: :tada:

**Review by vvrossem — APPROVED** [review]
Impressive one!

**Comment by vvrossem** [comment]
> The CI tests fail b/c:
> 
> `odoo.exceptions.UserError: ("You try to install module 'pos_tare' that depends on module 'base_fontawesome'.\nBut the latter module is not available in your system.", '')`
> 
> I'm not sure how to fix that!?

Neither do I, but I noticed the following logs appear in the console on the `pos.config` view:
```
Non loaded modules:       
Array [ "base_fontawesome5.FormRenderer", "base_fontawesome5.ListRenderer" ]
``` 
pos_config

File: pos_tare/static/src/js/screens.js (hunk 1 above)
Comment (Fkawala) [reply pos_tare/static/src/js/screens.js]: It is! I'll fix it.

File: pos_tare/static/src/js/screens.js (hunk 2 above)
Comment (Fkawala) [reply pos_tare/static/src/js/screens.js:132]: Kudos go to @legalsylvain on this (and most of the code actually).

File: pos_tare/static/src/js/screens.js (hunk 3 above)
Comment (Fkawala) [reply pos_tare/static/src/js/screens.js:158]: Glad you liked it! Thank you for the review :)

**Comment by Fkawala** [comment]
> > The CI tests fail b/c:
> > `odoo.exceptions.UserError: ("You try to install module 'pos_tare' that depends on module 'base_fontawesome'.\nBut the latter module is not available in your system.", '')`
> > I'm not sure how to fix that!?
> 
> Neither do I, but I noticed the following logs appear in the console on the `pos.config` view:
> 
> ```
> Non loaded modules:       
> Array [ "base_fontawesome5.FormRenderer", "base_fontawesome5.ListRenderer" ]
> ```
> 
> pos_config

Not sure how to read what you sent. The dependency is expected (we are trying to find a nice icon to make the tare information more explicit). The CI crash is kind of expected since it does not have the dependency (https://github.com/OCA/server-tools/tree/12.0/base_fontawesome) available.

**Comment by legalsylvain** [comment]
this [commit](https://github.com/OCA/pos/pull/491/commits/29c84d16bb08306f17a1d4ce846aa399d03a3552) should fix the problem. 
could you rebase ?

**Comment by Fkawala** [comment]
@legalsylvain thank you!

**Comment by Fkawala** [comment]
Howdy, can we try to close this one in what is left of this week?

**Comment by legalsylvain** [comment]
it seems you have some conflicts. could you try to fix it ? I guess a rebase should solve the issue.

**Comment by Fkawala** [comment]
Hello @legalsylvain, hope you're doing good! Am kind of lost with the codecov tool. Would you be able to provide any guidance to get that working? I looked around in codecov but https://codecov.io/gh/OCA/pos/tree/aa2ffbf91645ab059a36fe08b74a0f367b209cd0/pos_tare/models looks ok to me and I don't know what to do. Thank you!

**Comment by legalsylvain** [comment]
Hi @Fkawala. 
regarding codecov, it's not a blocking point. it just indicative to see if the coverage is good or not. for the current case, I don't understand why a 100% covered code make the ratio down...

I made some little improvment here : could you review it ? https://github.com/Fkawala/pos/pull/5

thanks !

File: pos_tare/static/src/js/screens.js (hunk 1 above)
Comment (legalsylvain) [reply pos_tare/static/src/js/screens.js]: done.

**Review by legalsylvain — APPROVED** [review]

**Comment by Fkawala** [comment]
Could we merge and close this PR? Could/Should I do it my self as the maintainer? Thank you and have a nice day!

**Comment by Fkawala** [comment]
@legalsylvain Merci!

**Pull request merged**